File: packages/autonavi_location_flutter/ios/Classes/AMapLocationAdapter.swift
import AMapLocationKit

/// Options passed to the location adapter. Plain Swift struct — no SDK types exposed.
struct LocationOptions {
    var accuracy: Int
    var intervalMs: Int
    var needAddress: Bool
}

/// Result returned from a successful location update. All SDK types are hidden.
typealias LocationResult = [String: Any?]

/// Adapter that isolates all AMapLocationKit calls.
/// When upgrading the AMap Location SDK, only this file needs changes.
class AMapLocationAdapter: NSObject, AMapLocationManagerDelegate {

    // MARK: - Callbacks (set by caller, no SDK types in signatures)

    var onLocation: ((LocationResult) -> Void)?
    var onError: ((code: String, message: String) -> Void)?

    // MARK: - Private

    private var manager: AMapLocationManager?
    private var isOnce = false
    private var onceCallback: ((LocationResult?, (code: String, message: String)?) -> Void)?

    // MARK: - Public API

    func startContinuous(options: LocationOptions) {
        isOnce = false
        let m = makeManager(options: options)
        m.startUpdatingLocation()
        manager = m
    }

    func startOnce(options: LocationOptions,
                   callback: @escaping (LocationResult?, (code: String, message: String)?) -> Void) {
        isOnce = true
        onceCallback = callback
        let m = makeManager(options: options)
        m.startUpdatingLocation()
        manager = m
    }

    func stop() {
        manager?.stopUpdatingLocation()
        manager?.delegate = nil
        manager = nil
    }

    func updateOptions(_ options: LocationOptions) {
        guard let m = manager else { return }
        applyOptions(options, to: m)
    }

    // MARK: - AMapLocationManagerDelegate

    func amapLocationManager(_ manager: AMapLocationManager!,
                             didUpdate location: CLLocation!,
                             reGeocode: AMapLocationReGeocode?) {
        let result = buildResult(location: location, reGeocode: reGeocode)
        if isOnce {
            let cb = onceCallback
            onceCallback = nil
            stop()
            cb?(result, nil)
        } else {
            onLocation?(result)
        }
    }

    func amapLocationManager(_ manager: AMapLocationManager!, didFailWithError error: Error!) {
        let nsError = error as NSError
        let err = (code: "LOCATION_ERROR_\(nsError.code)", message: nsError.localizedDescription)
        if isOnce {
            let cb = onceCallback
            onceCallback = nil
            stop()
            cb?(nil, err)
        } else {
            onError?(err)
        }
    }

    // MARK: - Private helpers

    private func makeManager(options: LocationOptions) -> AMapLocationManager {
        let m = AMapLocationManager()
        m.delegate = self
        applyOptions(options, to: m)
        return m
    }

    private func applyOptions(_ options: LocationOptions, to m: AMapLocationManager) {
        switch options.accuracy {
        case 0: m.desiredAccuracy = kCLLocationAccuracyHundredMeters
        case 1: m.desiredAccuracy = kCLLocationAccuracyNearestTenMeters
        case 3: m.desiredAccuracy = kCLLocationAccuracyBestForNavigation
        default: m.desiredAccuracy = kCLLocationAccuracyBest
        }
        m.locationTimeout = max(2, options.intervalMs / 1000)
        m.reGeocodeTimeout = max(2, options.intervalMs / 1000)
        m.locatingWithReGeocode = options.needAddress
    }

    private func buildResult(location: CLLocation, reGeocode: AMapLocationReGeocode?) -> LocationResult {
        return [
            "latitude": location.coordinate.latitude,
            "longitude": location.coordinate.longitude,
            "accuracy": location.horizontalAccuracy,
            "altitude": location.altitude,
            "speed": location.speed >= 0 ? location.speed : nil,
            "heading": location.course >= 0 ? location.course : nil,
            "timestamp": Int(location.timestamp.timeIntervalSince1970 * 1000),
            "address": reGeocode?.formattedAddress,
            "country": reGeocode?.country,
            "province": reGeocode?.province,
            "city": reGeocode?.city,
            "district": reGeocode?.district,
            "street": reGeocode?.street,
            "streetNum": reGeocode?.number,
            "cityCode": reGeocode?.citycode,
            "adCode": reGeocode?.adcode,
            "poiName": reGeocode?.poiName,
            "aoiName": reGeocode?.aoiName,
        ]
    }
}

File: packages/autonavi_location_flutter/android/src/main/kotlin/com/autonavi/location/flutter/AMapLocationAdapter.kt
package com.autonavi.location.flutter

import android.content.Context
import com.amap.api.location.AMapLocationClient
import com.amap.api.location.AMapLocationClientOption
import com.amap.api.location.AMapLocationClientOption.AMapLocationMode

/**
 * Options passed to the location adapter. Plain Kotlin data class — no SDK types exposed.
 */
data class LocationOptions(
    val accuracy: Int = 2,
    val intervalMs: Int = 2000,
    val needAddress: Boolean = true,
    val distanceFilter: Float = 0f,
)

// Top-level typealias — Kotlin does not allow typealias inside class bodies.
typealias LocationResult = Map<String, Any?>

/**
 * Adapter that isolates all AMapLocationClient calls.
 * When upgrading the AMap Location SDK, only this file needs changes.
 */
class AMapLocationAdapter(private val context: Context) {

    // Callbacks set by caller
    var onLocation: ((LocationResult) -> Unit)? = null
    var onError:    ((code: String, message: String?) -> Unit)? = null

    private var client: AMapLocationClient? = null

    // MARK: - Public API

    fun startContinuous(options: LocationOptions) {
        client = AMapLocationClient(context).apply {
            setLocationOption(buildOption(options))
            setLocationListener { result ->
                if (result.errorCode == 0) {
                    onLocation?.invoke(buildResultMap(result))
                } else {
                    onError?.invoke("LOCATION_ERROR_${result.errorCode}", result.errorInfo)
                }
            }
            startLocation()
        }
    }

    fun startOnce(
        options: LocationOptions,
        onSuccess: (LocationResult) -> Unit,
        onError: (code: String, message: String?) -> Unit,
    ) {
        val onceClient = AMapLocationClient(context)
        onceClient.setLocationOption(buildOption(options).apply { isOnceLocation = true })
        onceClient.setLocationListener { result ->
            onceClient.onDestroy()
            if (result.errorCode == 0) {
                onSuccess(buildResultMap(result))
            } else {
                onError("LOCATION_ERROR_${result.errorCode}", result.errorInfo)
            }
        }
        onceClient.startLocation()
    }

    fun stop() {
        client?.stopLocation()
        client?.onDestroy()
        client = null
    }

    fun updateOptions(options: LocationOptions) {
        client?.setLocationOption(buildOption(options))
    }

    // MARK: - Private helpers

    private fun buildOption(options: LocationOptions): AMapLocationClientOption {
        val mode = when (options.accuracy) {
            0    -> AMapLocationMode.Battery_Saving
            1    -> AMapLocationMode.Device_Sensors
            else -> AMapLocationMode.Hight_Accuracy
        }
        return AMapLocationClientOption().apply {
            locationMode = mode
            isNeedAddress = options.needAddress
            interval = options.intervalMs.toLong()
            deviceModeDistanceFilter = options.distanceFilter
        }
    }

    private fun buildResultMap(result: com.amap.api.location.AMapLocation): LocationResult =
        mapOf(
            "latitude"  to result.latitude,
            "longitude" to result.longitude,
            "accuracy"  to result.accuracy.toDouble(),
            "altitude"  to result.altitude,
            "speed"     to result.speed.toDouble(),
            "heading"   to result.bearing.toDouble(),
            "timestamp" to result.time,
            "address"   to result.address,
            "country"   to result.country,
            "province"  to result.province,
            "city"      to result.city,
            "district"  to result.district,
            "street"    to result.street,
            "streetNum" to result.streetNum,
            "cityCode"  to result.cityCode,
            "adCode"    to result.adCode,
            "poiName"   to result.poiName,
            "aoiName"   to result.aoiName,
        )
}
